### DISCUSSION TITLE
Using Multiple Sessions causes Pytest test to Hang

### DISCUSSION THREAD

**Discussion by mark-barrett** [sqlalchemy/sqlalchemy#9265] [Usage Questions]
We use Celery in our project and since it currently doesn't support asyncio, we have to use a sync session when we need to query the database in Celery and use and async session when we want to query it from our FastAPI app.

Having two sessions like this works fine in our app and we are able to query the database using both, however in our tests, it has always been an issue.

We have a fixture to yield an async session like this:

```python
@pytest.fixture()
async def db_session(db_engine) -> AsyncSession:
    async with db_engine.begin() as connection:
        await connection.begin_nested()
        await connection.run_sync(Base.metadata.drop_all)
        await connection.run_sync(Base.metadata.create_all)

        async with SessionLocal(bind=connection) as session:
            @event.listens_for(
                session.sync_session, "after_transaction_end"
            )
            def end_savepoint(session, transaction):
                if connection.closed:
                    return

                if not connection.in_nested_transaction():
                    connection.sync_connection.begin_nested()

            yield session
            await session.flush()
            await session.rollback()
```

This works fine and allows us to make asyncio queries in our tests.

I thought that perhaps, instead of having to create a new session for async, we could just use the `db_session.sync_session` on the async one, but when I do so in the tests, I get the usual:

```
sqlalchemy.exc.MissingGreenlet: greenlet_spawn has not been called; can't call await_() here. Was IO attempted in an unexpected place?
```

I create a sync version of the above async session to use in the tests, like this:

```python
@pytest.fixture()
def sync_db_session(sync_db_engine) -> Session:
    with sync_db_engine.begin() as connection:
        transaction = connection.begin_nested()
        Base.metadata.bind = connection
        Base.metadata.create_all()

        with Session(bind=connection, expire_on_commit=False, autoflush=True) as session:
            @event.listens_for(
                session, "after_transaction_end"
            )
            def end_savepoint(session, transaction):
                if connection.closed:
                    return

                if not connection.in_nested_transaction():
                    connection.begin_nested()

            yield session
            session.flush()
            session.rollback()
            transaction.rollback()
```

When I use just a single session in my tests like this:

```python
def test_some_test(self, db_session):
    # do stuff
    ...   
```

it works fine, but once I add both fixtures, like this:
```python
def test_some_test(self, db_session, sync_db_session):
    # do stuff
    ...   
```

the test will start, any asyncio calls to the db work, but once the test gets to a piece of code that calls the sync session e.g. `db.execute()` etc., it hangs until eventually it times out.

Is there something I'm missing here? Is the sync session waiting for a connection to the db or something? Is it advisable to have multiple sessions like this?

Any help is greatly apprecaited!

**Comment by CaselIT** [comment]
Hi,

you would need to ensure that you are not causing a deadlock on the db side, since these are two different connections.

in general there is nothing wrong with having multiple sessions, but care needs to be taken to ensure that they are not trying to modify the same object in two different transactions

**Reply by mark-barrett** [reply]
Thanks for the reply. I took a further look into it and tried different variations of the session to see if something is waiting.

It turns out when I comment out this in the Async session:

```python
await connection.run_sync(Base.metadata.drop_all)
await connection.run_sync(Base.metadata.create_all)
```

and comment this out in the Sync session:
```python
Base.metadata.bind = connection
Base.metadata.create_all()
```

and everything seems to work.

However, I'm then wondering how you go about cleaning the db between tests.

I put just one set of the drop/creates into it's own fixture, but low and behold it hangs again. Am I not able to make these calls when you have both sessions?

**Reply by zzzeek** [reply]
if you want to do drop/create DB for fixtures, that's also what we do for most tests, the trick is each different thread/process/whatever that's running tests concurrently needs to be referring to a different database entirely.     our test suite under multiprocessing creates new databases at startup and drops them at the end.

**Reply by CaselIT** [reply]
I'm assuming postgresql, since the other backend do not in general have a complete transactional ddl story.

What you are trying to do is to create the all the tables in the db on a transaction, so you cannot access any table from another one until the first transaction is committed. Note that this behaviour comes from SQL and the db, not from sqlalchemy. You can find information about how postgresql behaves here https://www.postgresql.org/docs/current/tutorial-transactions.html

**Reply by mark-barrett** [reply]
What I'm trying to do is clear the contents of each table for each test. Is this the most advisable way to do that, by dropping and creating all the tables? 

@CaselIT so for the calls below:

```python
await connection.run_sync(Base.metadata.drop_all)
await connection.run_sync(Base.metadata.create_all)
```

are you saying that these run and the other session tries to access the tables but because they haven't been committed it hangs?

Do I need to manually commit after issuing these commands?

If the answer is yes, I assume the sync fixture doesn't need to do the same? Or should those calls be put in anyway?

**Reply by mark-barrett** [reply]
Found this fixture online, seems to do the trick:

```python
@pytest.fixture(autouse=True)
def clean_db(sync_db_engine):
    connection = sync_db_engine.connect()
    transaction = connection.begin()
    for name, table in Base.metadata.tables.items():
        connection.execute(table.delete())
    transaction.commit()
```

is that advisable instead of calling the `Base.metadata.drop_all`

**Reply by zzzeek** [reply]
clearing the tables is good also, if all your tests run with the same schema, then yes definitely clearing tables is much more lightweight.

still doesn't protect you from concurrency situations, if you have two tests working with the same table at once, that won't work.

**Reply by CaselIT** [reply]
Yes, if the schema does not change, delete is definitely faster than dropping and recreating the schema.

You may also use `truncate` instead of delete, that should be faster still.
